### DISCUSSION TITLE
Using unit 'week' requires that a relative reference time be given, but none was provided

### DISCUSSION THREAD

**Discussion by azzamsa** [BurntSushi/jiff#84] [Q&A] [answered]
Hi, 👋🏽 

Why this is the case?
Why the chrono code works but jiff doesn't.

```rust
    let dt = 1.weeks();
    dbg!(&dt);
    let dt = dt.total(Unit::Second);
    dbg!(&dt);

    let dt = chrono::Duration::weeks(1);
    dbg!(&dt);
    let dt = dt.num_seconds();
    dbg!(&dt);
```

```rust
[src/main.rs:21:5] &dt = P1w
[src/main.rs:23:5] &dt = Err(
    Error {
        inner: ErrorInner {
            kind: Adhoc(
                using unit 'week' requires that a relative reference time be given, but none was provided,
            ),
            cause: None,
        },
    },
)
[src/main.rs:26:5] &dt = TimeDelta {
    secs: 604800,
    nanos: 0,
}
[src/main.rs:28:5] &dt = 604800
```

**Comment by azzamsa** [comment]
>  Jiff specifically avoids returning "unbalanced" spans like 2 months 35 days. So add a date to the mix and things work fine:

Explained [here](https://github.com/BurntSushi/jiff/discussions/41#discussioncomment-10141487)

**Reply by BurntSushi** [reply, marked as answer]
Yeah basically you need to have a relative date to do transformations like this. Jiff is preventing you from introducing bugs in subtle corner cases. Indeed, `chrono-humanize` [just assumes that all months are 30 days](https://gitlab.com/imp/chrono-humanize-rs/-/blob/master/src/humantime.rs?ref_type=heads#L46), which is obviously wrong:

```rust
use chrono_humanize::{Accuracy, HumanTime, Tense};

fn main() {
    let dt = chrono::Duration::days(31);
    let ht = HumanTime::from(dt);
    let text = ht.to_text_en(Accuracy::Precise, Tense::Present);
    println!("{text}");
}
```

Has this output:

```
1 month and 1 day
```

Some months are 31 days. Some are 29. Some are 28. Some are 30. The only correct way to know the duration of a month is to interpret it relative to a particular date.

**Comment by azzamsa** [comment]
I have successfully ported `chrono-humanize` to jiff. But I still disable some tests due to this.

https://github.com/azzamsa/jiffy/commit/a16f8c6ab2ece23296ba2bd6befe238108e53e5b

**Reply by BurntSushi** [reply]
It looks like you've [copied `chrono-humanize`'s mistake](https://github.com/azzamsa/jiffy/blob/5f3c539378eea2de542bb5a8f977d2704cc8755f/src/humantime.rs#L51). :-( I would strongly urge you not to do that. Jiff expends _so much effort_ in its API design to prevent folks from making that exact mistake.

**Reply by azzamsa** [reply]
> It looks like you've [copied `chrono-humanize`'s mistake](https://github.com/azzamsa/jiffy/blob/5f3c539378eea2de542bb5a8f977d2704cc8755f/src/humantime.rs#L51). :-( I would strongly urge you not to do that. Jiff expends _so much effort_ in its API design to prevent folks from making that exact mistake.

You're right. Just after you mentioned this:

> Indeed, chrono-humanize [just assumes that all months are 30 days](https://gitlab.com/imp/chrono-humanize-rs/-/blob/master/src/humantime.rs?ref_type=heads#L46), which is obviously wrong:

I started thinking:
- I'll ask for your input on any corner cases I should be aware of.
- I'll simply use all of `jiff`'s functions like `get_days`, `get_months`, etc. This way, it's just a thin wrapper, and `jiff` does all the heavy lifting. `Jiff` already provides everything I need, so there's no need to compute anything on my own.

**Reply by BurntSushi** [reply]
> I'll simply use all of jiff's functions like get_days, get_months, etc. This way, it's just a thin wrapper, and jiff does all the heavy lifting. Jiff already provides everything I need, so there's no need to compute anything on my own.

Right! That is exactly the benefit of `Span`. It is not the normal "Integer number of nanoseconds" that you are probably used to. It keeps track of individual units. It was designed by the Temporal project to be compatible with RFC 5545, and so that any code for localizing the `Span` wouldn't actually need to do any duration math or make choices about which units to use.
